src/redisearch_rs/rqe_iterators/src/optional.rs

```diff
@@ -0,0 +1,224 @@
+            .child
+            .as_mut()
+            .map(|child| {
+                let child_last_doc_id = child.last_doc_id();
+                match child_last_doc_id.cmp(&(self.result.doc_id + 1)) {
+                    cmp::Ordering::Less => child.read(),
+                    cmp::Ordering::Equal => Ok(child.current()),
+                    cmp::Ordering::Greater => Ok(None),
+                }
+            })
+            .transpose()?
+            .flatten();
+
+        self.result.doc_id += 1;
+
+        if let Some(real) = maybe_real {
+            debug_assert!(
+                real.doc_id >= self.result.doc_id,
+                "no backwards reads should be possible"
+            );
+
+            if real.doc_id > self.result.doc_id {
+                // demote to virtual, preserve sequential guarantees
+                return Ok(Some(&mut self.result));
+            }
+
+            real.weight = self.weight;
+            return Ok(Some(real));
+        }
+
+        Ok(Some(&mut self.result))
+    }
+
+    // C-Code: SkipTo for OPTIONAL iterator - Non-optimized version.
+    // Skip to a specific docId. If the child has a hit on this docId, return it.
+    // Otherwise, return a virtual hit.
+    fn skip_to(
+        &mut self,
+        doc_id: t_docId,
+    ) -> Result<Option<SkipToOutcome<'_, 'index>>, RQEIteratorError> {
+        debug_assert!(doc_id > self.result.doc_id);
+
+        if doc_id > self.max_doc_id || self.at_eof() {
+            self.result.doc_id = self.max_doc_id;
+            return Ok(None);
+        }
+
+        if let Some(child) = self.child.as_mut() {
+            if doc_id > child.last_doc_id() {
+                // use current() here to work around
+                // borrowing rules to be able to handle
+                // both of `doc_id >= child.last_doc_id` cases...
+                let _ = child.skip_to(doc_id)?;
+            }
+
+            if let Some(real) = child.current()
+                && real.doc_id == doc_id
+            {
+                real.weight = self.weight;
+                self.result.doc_id = real.doc_id;
+                return Ok(Some(SkipToOutcome::Found(real)));
+            }
+        }
+
+        self.result.doc_id = doc_id;
+        Ok(Some(SkipToOutcome::Found(&mut self.result)))
+    }
+
+    fn revalidate(&mut self) -> Result<RQEValidateStatus<'_, 'index>, RQEIteratorError> {
+        let Some(ref mut child) = self.child else {
+            return Ok(RQEValidateStatus::Ok);
+        };
+        let last_child_doc_id = child.last_doc_id();
+
+        // Revalidate the child iterator (C-Code: step 1)
+        match child.revalidate()? {
+            // Abort: Handle child validation results (but continue processing)
+            // C-Code: step 2
+            RQEValidateStatus::Aborted => {
+                self.child = None; // Drop it so we become fully virtual until max is reached
+                Ok(if last_child_doc_id != self.result.doc_id {
```

Comment:
Referring to https://github.com/RediSearch/RediSearch/pull/7046/files#r2596406983,
Could there be a case where this check falsely returns `true`/`false`?